core/src/main/java/ai/timefold/solver/core/api/domain/metamodel/DefaultUnassignedLocation.java

```diff
@@ -0,0 +1,11 @@
+package ai.timefold.solver.core.api.domain.metamodel;
+
+record DefaultUnassignedLocation() implements UnassignedLocation {
```

Comment:
I understand your strategy, but I suggest a different approach. What if `DefaultUnassignedLocation` extends `DefaultLocationInList` and throws exceptions for `DefaultUnassignedLocation::entity()` or `DefaultUnassignedLocation::index()`?

We still fail fast, and there will be no need to cast values in the internal classes, which I didn't like when using `LocationInList` previously. Additionally, we could add a method `ElementLocation::isAssigned()` and override it in `DefaultUnassignedLocation`.